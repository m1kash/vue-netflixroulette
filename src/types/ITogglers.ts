interface IToggleElems {
  name: string,
  key: string | number
}

export default IToggleElems;
