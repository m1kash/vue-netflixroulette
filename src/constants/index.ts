import ITogglers from "@/types/ITogglers";

export const FILTER_PARAMS_SEARCH = [
  {
    name: 'title',
    id: 'title',
    key: '12'
  },
  {
    name: 'genre',
    id: 'genres',
    key: '21'
  }
];
export const FILTER_PARAMS_TOOLBAR = [
  {
    name: 'release date',
    id: 'release_date',
    key: '12'
  },
  {
    name: 'rating',
    id: 'vote_average',
    key: '21'
  }
] as Array<ITogglers>;
