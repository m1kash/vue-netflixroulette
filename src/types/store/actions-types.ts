import { ActionContext, ActionTree } from 'vuex';
import IState from '@/types/IState';
import { searchBy } from '@/types/searchTag';
import { sortBy } from '@/types/sortBy';

export enum ActionsTypes {
  FILTER_SEARCH_TEXT = 'FILTER_SEARCH_TEXT',
  FILTER_SEARCH_BY = 'FILTER_SEARCH_BY',
  CHANGE_SORT_BY = 'CHANGE_SORT_BY',
  GET_MOVIE_BY_ID = 'GET_MOVIE_BY_ID',
}

export interface IActions extends ActionTree<IState, IState> {
  [ActionsTypes.FILTER_SEARCH_TEXT](
    context: ActionContext<IState, IState>,
    payload: string):void
  [ActionsTypes.FILTER_SEARCH_BY](
    context: ActionContext<IState, IState>,
    payload: searchBy):void
  [ActionsTypes.CHANGE_SORT_BY](
    context: ActionContext<IState, IState>,
    payload: sortBy):void
  [ActionsTypes.GET_MOVIE_BY_ID](
    context: ActionContext<IState, IState>,
    payload: string):void
}
