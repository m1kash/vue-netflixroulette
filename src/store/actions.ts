import {TActions, ActionsTypes} from "@/types/store/actions-types";
import {MutationTypes} from "@/types/store/mutation-types";

export const actions: TActions = {
  [ActionsTypes.FILTER_SEARCH_TEXT]({ commit }, payload) {
    commit(MutationTypes.SET_SEARCH_TEXT, payload);
    commit(MutationTypes.FILTER_MOVIES);
  },
  [ActionsTypes.FILTER_SEARCH_BY]({ commit }, payload) {
    commit(MutationTypes.SET_SEARCH_BY, payload);
    commit(MutationTypes.FILTER_MOVIES);
  },
  [ActionsTypes.CHANGE_SORT_BY]({ commit }, payload) {
    commit(MutationTypes.SET_SORT_BY, payload);
    commit(MutationTypes.FILTER_MOVIES);
  },
  [ActionsTypes.GET_MOVIE_BY_ID]({ commit }, payload) {
    commit(MutationTypes.SET_MOVIE, payload);
    commit(MutationTypes.FILTER_RELATED_MOVIES);
  },
}
