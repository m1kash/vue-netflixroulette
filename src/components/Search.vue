<template>
  <div class="flex flex-wrap w-full h-full p-20">
    <span class="uppercase font-extralight text-4xl text-white w-full mb-4"
     for="search">
    Find Your Movie
    </span>
    <form class="space-x-4 flex w-full aqa-form-search" @submit.prevent="search">
      <text-field v-model:value="value" @keyup="search"
                  width="w-3/4" id="search"
                  placeholder="Search" />
      <large-button width="w-1/4" @click.prevent="search">
        Search
      </large-button>
    </form>
    <portal-toggle label="Search by" :active="activeFilter"
                   :toggle-elems="elems"
                   @change-toggler="changeActiveElem"></portal-toggle>
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue';
import TextField from '@/components/TextField.vue';
import LargeButton from '@/components/LargeButton.vue';
import portalToggle from '@/components/Toggle.vue';
import { ActionsTypes } from '@/types/store/actions-types';
import ITogglers from '@/types/ITogglers';
import { searchBy } from '@/types/searchTag';
import { useStore } from '@/store';

export default defineComponent({
  name: 'portal-search',
  data: () => ({
    elems: [] as ITogglers[],
    value: '' as string,
    activeFilter: '' as string,
  }),
  props: {
    togglers: {
      type: Array as PropType<ITogglers[]>,
      required: true,
    },
  },
  mounted() {
    const store = useStore();

    this.elems = this.$props.togglers ? this.$props.togglers : [];
    this.activeFilter = store.state.searchBy;
  },
  methods: {
    search() {
      let config = {};
      this.$store.dispatch(ActionsTypes.FILTER_SEARCH_TEXT, this.value);
      if (this.value) {
        config = { query: { search: this.value } };
      }
      this.$router.push({ name: 'Home', ...config });
    },
    changeActiveElem(id: searchBy) {
      this.$store.dispatch(ActionsTypes.FILTER_SEARCH_BY, id);
    },
  },
  components: { portalToggle, LargeButton, TextField },
});
</script>
