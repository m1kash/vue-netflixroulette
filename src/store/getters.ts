import { GetterTree } from 'vuex';
import IState from '@/types/IState';
import { Getters } from '@/types/store/getters-types';

const getters: GetterTree<IState, IState> & Getters = {
  getMovies(state) {
    return state.items;
  },
  getMovie(state) {
    return state.movie;
  },
  getRelatedMovies(state) {
    return state.relatedMovies;
  },
  getCountMovies(state) {
    return state.items.length;
  },
  getFirstGenreMovie(state) {
    return state.movie.genres ? state.movie.genres[0] : '';
  },
};

export default getters;
