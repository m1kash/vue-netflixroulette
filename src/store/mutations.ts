import { MutationTree } from 'vuex';
import IState from '@/types/IState';
import { Mutations, MutationTypes } from '@/types/store/mutation-types';

const mutations: MutationTree<IState> & Mutations = {
  [MutationTypes.SET_MOVIE](state, payload) {
    state.movie = payload;
  },
  [MutationTypes.SET_SEARCH_TEXT](state, payload) {
    state.searchText = payload;
  },
  [MutationTypes.SET_SEARCH_BY](state, payload) {
    state.searchBy = payload;
  },
  [MutationTypes.SET_SORT_BY](state, payload) {
    state.sortBy = payload;
  },
  [MutationTypes.SET_MOVIES](state, payload) {
    state.items = payload;
  },
  [MutationTypes.FILTER_RELATED_MOVIES](state, payload) {
    state.relatedMovies = payload;
  },
};

export default mutations;
