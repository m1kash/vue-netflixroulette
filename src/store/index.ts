import IState from '@/types/IState';
import {createStore, Store, useStore as baseUseStore} from 'vuex'
import movies from '../../data/movies.json';
import { getters } from "@/store/getters";
import {mutations} from "@/store/mutations";
import {actions} from "@/store/actions";
import {InjectionKey} from "vue";

export const key: InjectionKey<Store<IState>> = Symbol();
const store = createStore<IState>({
  state () {
    return {
      items: movies,
      filteredMovies: [],
      relatedMovies: [],
      movie: {},
      sortBy: 'release_date',
      searchBy: 'title',
      searchText: ''
    }
  },
  getters: getters,
  mutations: mutations,
  actions: actions
});

export function useStore() {
  return baseUseStore(key);
}

export default store;
