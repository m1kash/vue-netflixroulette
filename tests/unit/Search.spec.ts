import 'jest';
import { shallowMount, mount, VueWrapper } from '@vue/test-utils';
import Search from '@/components/Search.vue';
import { createStore } from 'vuex';
import { ActionsTypes } from '@/types/store/actions-types';
import { key } from '@/store';
import {FILTER_PARAMS_SEARCH} from "@/constants";
import ITogglers from "@/types/ITogglers";

describe('Search.vue', () => {
  let wrapper: VueWrapper;
  let store;
  let filterSearchMock: jest.Mock;

  beforeEach(() => {
    filterSearchMock = jest.fn();
    store = createStore({
      actions: {
        [ActionsTypes.FILTER_SEARCH_TEXT]: filterSearchMock,
      }
    });
    wrapper = mount(Search, {
      global: {
        provide: {
          [key]: store,
        },
        plugins: [
          store
        ]
      },
      props: {
        togglers: FILTER_PARAMS_SEARCH as Array<ITogglers>,
      },
      active: 'genres',
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('render correct', () => {
    expect(wrapper.vm.$el).toMatchSnapshot();
  });

  test('user can search movie and pressing enter or button', async () => {
    const form = wrapper.find('.aqa-form-search');
    const button = wrapper.findComponent({ name: 'large-button' });
    const field = wrapper.findComponent({ name: 'text-field' });

    wrapper.find('#search-input').setValue('Movie name');
    button.trigger('click.prevent');
    await wrapper.vm.$nextTick();

    expect(filterSearchMock).toBeCalledWith(expect.anything(), 'Movie name');

    filterSearchMock.mockClear();

    wrapper.find('#search-input').setValue('Movie name 2');
    form.trigger('submit.prevent');

    await wrapper.vm.$nextTick();

    expect(filterSearchMock).toBeCalledWith(expect.anything(), 'Movie name 2');
  });
});
