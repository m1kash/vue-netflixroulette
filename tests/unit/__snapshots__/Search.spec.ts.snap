// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search.vue render correct 1`] = `
<div
  class="flex flex-wrap w-full h-full p-20"
>
  <span
    class="uppercase font-extralight text-4xl text-white w-full mb-4"
    for="search"
  >
     Find Your Movie 
  </span>
  <form
    class="space-x-4 flex w-full aqa-form-search"
  >
    <input
      class="w-3/4 placeholder:text-stone-500 text-lg bg-stone-800 px-4 py-2 opacity-75 text-white w-3/4 h-12 border-0 rounded"
      id="search-input"
      placeholder="Search"
      type="text"
    />
    <button
      class="w-1/4 bg-rose-500 h-12 text-white rounded uppercase size text-lg hover:bg-rose-600"
    >
      
       Search 
      
    </button>
  </form>
  <div
    class="flex items-center py-4"
  >
    <span
      class="text-white uppercase text-lg mr-4"
    >
      Search by
    </span>
    <div
      class="flex h-full"
    >
      
      <button
        class="toggle-btn rounded-l"
      >
        title
      </button>
      <button
        class="toggle-btn rounded-r"
      >
        genre
      </button>
      
    </div>
  </div>
</div>
`;
